File: src/02_Time_complexity/max_subarray_sum.cpp
#include<bits/stdc++.h>
#include "../compete.hpp"
using namespace std;

vector<int> v{-1, 2, 4, -3, 5, 2, -5, 2};


// O(n)
void solve()
{
  int sum = 0;
  int best = -1e9;
  for (auto i : v) {
    sum = max(sum + i, i);
    best = max(best, sum);
  }
  cout << best << '\n';
}


// O(n)
void max_subarray()
{
  int sum = 0;
  int best = -1e9;
  vector<int> sub;
  vector<int> best_sub;

  for (int& x : v) {
    sum += x;
    if (x > sum) {
      sum = x;
      sub.clear();
      sub.push_back(x);
    } else sub.push_back(x);
    // cout << setw(2) << x << ' '
    //      << setw(2) << sum << ' ';
    // cout << sub << '\n';

    if (sum > best) {
      best = sum;
      best_sub = sub;
    }
  }

  cout << best << '\n';
  cout << best_sub << '\n';
}

int main() {max_subarray();}

File: src/02_Time_complexity/max_subarray_sum.py
v = [-1, 2, 4, -3, 5, 2, -5, 2]


# O(n^2)
def solve_1():
  if not v:
    return 0
  if len(v) == 1:
    return max(0, v[0])

  best = 0
  for i in range(len(v)):
    s = 0
    for j in range(i, len(v)):
      s += v[j]
      if s > best:
        best = s
  return best


# O(n)
# Kadane's alg
def solve_2():
  if not v:
    return 0
  if len(v) == 1:
    return max(0, v[0])

  best = 0
  s = 0

  for x in v:
    s += x
    if x > s:
      s = x
    if s > best:
      best = s

  return best


def max_sum_subarray():
  if not v:
    return []
  if len(v) == 1:
    if v[0] < 0:
      return []
    return v

  best = 0
  s = 0
  best_sub = []
  sub = []

  for x in v:
    s += x
    if x > s:
      s = x
      sub = [x]
    else:
      sub.append(x)
    if s > best:
      best = s
      best_sub = sub.copy()

  return best_sub


if __name__ == '__main__':
  print(v)
  print("max subarray sum:", solve_2())
  print("max sum subarray:", max_sum_subarray())
